core/templates/pages/exploration-editor-page/editor-tab/services/responses.service.spec.ts

```diff
@@ -0,0 +1,678 @@
+
+    var expectedAnswerGroup = interactionDataWithRules.answerGroups;
+    expectedAnswerGroup[0].rules[0].inputs.x = ['f', 'd', 'e'];
+    expectedAnswerGroup[0].rules[0].inputs.y = ['d', 'e', 'f'];
+
+    expect(callbackSpy).toHaveBeenCalledWith(expectedAnswerGroup);
+    expect(ResponsesService.getAnswerGroup(0)).toEqual(
+      expectedAnswerGroup[0]);
+    expect(ResponsesService.getAnswerChoices()).toEqual(newAnswerChoices);
+  });
+
+  it('should update answer choices when savedMemento is' +
+    ' DragAndDropSortInput and rule type is' +
+    ' HasElementXAtPositionY', function() {
+    interactionDataWithRules.id = 'DragAndDropSortInput';
+    interactionDataWithRules.answerGroups[0].rules[0].type =
+      'HasElementXAtPositionY';
+    interactionDataWithRules.answerGroups[0].rules[0].inputs.y = [];
+
+    ResponsesService.init(interactionDataWithRules);
+    StateEditorService.setInteraction(interactionDataWithRules);
+    StateInteractionIdService.init('stateName', 'DragAndDropSortInput');
+
+    ResponsesService.updateAnswerChoices([{
+      val: 'a'
+    }, {
+      val: 'b'
+    }, {
+      val: 'c'
+    }], function() {});
+
+    var newAnswerChoices = [{
+      val: 'c'
+    }, {
+      val: 'b'
+    }];
+    var callbackSpy = jasmine.createSpy('callback');
+    ResponsesService.updateAnswerChoices(newAnswerChoices, callbackSpy);
+
+    var expectedAnswerGroup = interactionDataWithRules.answerGroups;
+    expectedAnswerGroup[0].rules[0].inputs.x = '';
+    expectedAnswerGroup[0].rules[0].inputs.y = 1;
+
+    expect(callbackSpy).toHaveBeenCalledWith(expectedAnswerGroup);
+    expect(ResponsesService.getAnswerChoices()).toEqual(newAnswerChoices);
+  });
+
+  it('should update answer choices when savedMemento is' +
+    ' DragAndDropSortInput and rule type is' +
+    ' HasElementXBeforeElementY', function() {
+    interactionDataWithRules.id = 'DragAndDropSortInput';
+    interactionDataWithRules.answerGroups[0].rules[0].type =
+      'HasElementXBeforeElementY';
+
+    ResponsesService.init(interactionDataWithRules);
+    StateEditorService.setInteraction(interactionDataWithRules);
+
+    StateInteractionIdService.init('stateName', 'DragAndDropSortInput');
+
+    ResponsesService.updateAnswerChoices([{
+      val: 'a'
+    }, {
+      val: 'b'
+    }, {
+      val: 'c'
+    }], function() {});
+
+    var newAnswerChoices = [{
+      val: 'd'
+    }, {
+      val: 'e'
+    }, {
+      val: 'f'
+    }];
+    var callbackSpy = jasmine.createSpy('callback');
+    ResponsesService.updateAnswerChoices(newAnswerChoices, callbackSpy);
+
+    var expectedAnswerGroup = interactionDataWithRules.answerGroups;
+    expectedAnswerGroup[0].rules[0].inputs.x = '';
+    expectedAnswerGroup[0].rules[0].inputs.y = '';
+
+    expect(callbackSpy).toHaveBeenCalledWith(expectedAnswerGroup);
+    expect(ResponsesService.getAnswerChoices()).toEqual(newAnswerChoices);
+  });
+
+  it('should update answer choices when savedMemento is' +
```

Comment:
I don't understand this test...

Reply:
I've explained it in comment above. PTAL! Let me know if it's still confusing.